[ImportExport]ARM REST API

RenaShahMSFT opened the pull request:
This checklist is used to make sure that common issues in a pull request are addressed. This will expedite the process of getting your pull request merged and avoid extra work on your part to fix issues discovered during the review process. 

### PR information
- [ ] The title of the PR is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For information on cleaning up the commits in your pull request, [see this page](https://github.com/Azure/azure-powershell/blob/dev/documentation/cleaning-up-commits.md).
- [ ] Except for special cases involving multiple contributors, the PR is started from a fork of the main repository, not a branch.
- [ ] If applicable, the PR references the bug/issue that it fixes.
- [ ] Swagger files are correctly named (e.g. the `api-version` in the path should match the `api-version` in the spec).

### Quality of Swagger
- [ ] I have read the [contribution guidelines](https://github.com/Azure/azure-rest-api-specs/blob/master/.github/CONTRIBUTING.md).
- [ ] My spec meets the review criteria:
  - [ ] The spec conforms to the [Swagger 2.0 specification](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md).
  - [ ] Validation errors from the [Linter extension for VS Code](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/linter.md) have all been fixed for this spec. (**Note:** for large, previously checked in specs, there will likely be many errors shown. Please contact our team so we can set a timeframe for fixing these errors if your PR is not going to address them).
  - [ ] The spec follows the patterns described in the [Swagger good patterns](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-good-patterns.md) document unless the service API makes this impossible.
New submission for REST API documentation

RenaShahMSFT committed:
[ImportExport]ARM REST API

New submission for REST API documentation

azurecla commented:
Hi __@RenaShahMSFT__, I'm your friendly neighborhood Azure Pull Request Bot (You can call me AZPRBOT). Thanks for your contribution!
    <p>
        It looks like you're working at Microsoft (renash). If you're full-time, we DON'T require a contribution license agreement.
    </p>
    <p>
        If you are a vendor, DO please sign the electronic contribution license agreement. It will take 2 minutes and there's no faxing! https://cla.azure.com.
    </p>

TTYL, AZPRBOT;

amarzavery commented:
do not review the PR. I have emailed a copy of the updated spec after making changes locally in vscode.

amarzavery commented:
@RenaShahMSFT - The checkmarks in the PR description are for a reason. Please make sure that you have gone through them next time while submitting the PR.

RenaShahMSFT committed:
Addressing comments

Thanks for the detailed feedback Amar. Resolved the comments in attached file.
Rena please upload both versions so people can see the diffs.

Thanks,
Chun

amarzavery commented on arm-storageimportexport/2016-11-01/swagger/storageimportexport.json:
```diff
@@ -0,0 +1,1194 @@
+          {
+            "name": "$filter",
+            "in": "query",
+            "description": "Can be used to restrict the results to certain conditions. The following possible values can be used with $filter: 1) $filter=type eq '{type}'; 2) $filter=trackingnumber eq '{trackingnumber}'; 3) $filter=state eq '{state}'; 4) Logical and combination of the above, for example: $filter=type eq 'Import' and state eq 'Transferring'. Valid values for type are Import and Ex ...
```
From this comment it looks like the "state" property has a set of valid values. This makes state to be an enum candidate. Can you please add the enum array and the x-ms-enum extension to the state property. Btw, I found MutableJobProperties, JobProperties and DriveStatus all of these models have a state property.  So the state belongs to a Job or DriveStatus?

mmacy commented on arm-storageimportexport/2016-11-01/swagger/storageimportexport.json:
```diff
@@ -0,0 +1,1194 @@
+      ]
+    },
+    "ErrorResponse": {
+      "description": "It describes the model for Error Response",
```
Strike "It" from this sentence.

mmacy commented on arm-storageimportexport/2016-11-01/swagger/storageimportexport.json:
```diff
@@ -0,0 +1,1194 @@
+          "$ref": "#/definitions/ReturnAddress"
+        },
+        "returnShipping": {
+          "description": "Specifies the return carrier and customer’s account with the carrier.",
```
Use straight quote: customer's

mmacy commented on arm-storageimportexport/2016-11-01/swagger/storageimportexport.json:
```diff
@@ -0,0 +1,1194 @@
+          "type": "object"
+        },
+        "properties": {
+          "description": "Specifies the properties of a UpdateJob.",
```
an UpdateJob

mmacy commented on arm-storageimportexport/2016-11-01/swagger/storageimportexport.json:
```diff
@@ -0,0 +1,1194 @@
+        },
+        "containerSas": {
+          "type": "string",
+          "description": "The container SAS to be used to import or export data to or from the storage account."
```
The container shared access signature (SAS) to be used...

mmacy commented on arm-storageimportexport/2016-11-01/swagger/storageimportexport.json:
```diff
@@ -0,0 +1,1194 @@
+      }
+    },
+    "Job": {
+      "description": "Describes an import/export job",
```
Missing terminating period.

mmacy commented on arm-storageimportexport/2016-11-01/swagger/storageimportexport.json:
```diff
@@ -0,0 +1,1194 @@
+      ],
+      "properties": {
+        "properties": {
+          "description": "Import/export job properties",
```
Missing terminating period.

mmacy commented on arm-storageimportexport/2016-11-01/swagger/storageimportexport.json:
```diff
@@ -0,0 +1,1194 @@
+      ]
+    },
+    "OperationDisplayInformation": {
+      "description": "Display Information about the operation.",
```
Display information

mmacy commented on arm-storageimportexport/2016-11-01/swagger/storageimportexport.json:
```diff
@@ -0,0 +1,1194 @@
+          "type": "string"
+        },
+        "description": {
+          "description": "Short Description of the operation.",
```
Short description

mmacy commented on arm-storageimportexport/2016-11-01/swagger/storageimportexport.json:
```diff
@@ -0,0 +1,1194 @@
+      "required": ["name", "display"],
+      "properties": {
+        "name": {
+          "description": "name of the operation.",
```
Name of the operation.

mmacy commented on arm-storageimportexport/2016-11-01/swagger/storageimportexport.json:
```diff
@@ -0,0 +1,1194 @@
+          "type": "string"
+        },
+        "display": {
+          "description": "Display Information about the operation.",
```
Display information

mmacy commented on arm-storageimportexport/2016-11-01/swagger/storageimportexport.json:
```diff
@@ -0,0 +1,1194 @@
+          "type": "string"
+        },
+        "properties": {
+          "description": "Describes the properties of a Location.",
```
...properties of a location.

mmacy commented on arm-storageimportexport/2016-11-01/swagger/storageimportexport.json:
```diff
@@ -0,0 +1,1194 @@
+        },
+        "phone": {
+          "type": "string",
+          "description": "Phone number of the recipient of the returned drives. Is there any requirement for the country code? Is there any specific format that the server expects?"
```
I think this description includes a TODO from Amar's review (just missing the TODO):

 Is there any requirement for the country code? Is there any specific format that the server expects?

mmacy commented on arm-storageimportexport/2016-11-01/swagger/storageimportexport.json:
```diff
@@ -0,0 +1,1194 @@
+      }
+    },
+    "DriveStatus": {
+      "description": "Provides information about the drives status.",
```
drive's

mmacy commented on arm-storageimportexport/2016-11-01/swagger/storageimportexport.json:
```diff
@@ -0,0 +1,1194 @@
+        },
+        "blobListblobPath": {
+          "type": "string",
+          "description": "The relative URI to the block blob that contains the list of blob paths or blob path prefixes as defined above, beginning with the container name. If the blob is in root container, the URI must begin with $root."
```
If the blob is in the root container

mmacy commented on arm-storageimportexport/2016-11-01/swagger/storageimportexport.json:
```diff
@@ -0,0 +1,1194 @@
+      "properties": {
+        "targetResourceGroup": {
+          "type": "string",
+          "description": "Specifies the target resource group id to move the jobs to."
```
resource group ID to

mmacy reviewed (commented):
General documentation pass.

RenaShahMSFT committed:
Addrssing comments from Amar and Marsh

RenaShahMSFT committed:
Updating the Swagger file with comments addressed

mmacy committed:
Doc edits for import/export Swagger spec

amarzavery commented on arm-storageimportexport/2016-11-01/swagger/storageimportexport.json:
```diff
@@ -0,0 +1,1209 @@
+        "tags": [
+          "StorageImportExport"
+        ],
+        "operationId": "Jobs_ListBySubscription",
```
I am sorry if I named it like that. As per the convention this should be named as `Jobs_List`

amarzavery replied:
@RenaShahMSFT - Can you please make this last update?

cyuanmsft replied:
Operations ids we had before review:
"operationId": "ListLocations",
"operationId": "GetLocation",
"operationId": "ListJobsInSubscription",
"operationId": "ListJobsInResourceGroup",
"operationId": "GetJob",
"operationId": "UpdateJob",
"operationId": "PutJob",
"operationId": "DeleteJob",
"operationId": "Movejobs",
"operationId": "GetBitLockerKeys",
"operationId": "ListAvailableOperations",

After review:
"operationId": "ListLocations",
"operationId": "GetLocation",
"operationId": "Jobs_ListBySubscription",
"operationId": "Jobs_ListByResourceGroup",
"operationId": "Jobs_Get",
"operationId": "Jobs_Update",
"operationId": "Jobs_CreateOrUpdate",
"operationId": "Jobs_Delete",
"operationId": "Jobs_Move",
"operationId": "Jobs_ListBitLockerKeys",
"operationId": "ListSupportedOperations",

1) The change doesn't look consistent.
2) List Jobs can be per Resource Group vs. per Subscription, hence the need to distinguish them.

amarzavery replied:
The convention that we have chosen is to use the name List for listing resources in a subscription and listByResourceGroup for listing resources in a resource group. Please take a look at the recommended patterns over [here](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-good-patterns.md#swagger-recommended-patterns).

This in particular

Consistency in the naming of list methods
Operations that provide resources in a resource group should be named as "OperationGroup_ListByResourceGroup"
Operations that provide resources in a subscription should be named as "OperationGroup_List"
If the ListByResourceGroup is overloaded "i.e. if the resourcegroup is null then the operation lists all resources in the subscription" then, such operations should not be allowed.
If the list operation is on a child resource type then please follow the convention of Noun_ListBy{Parent}. For example "Databases" is a child resource type of "SqlServer". So for listing all the databases the operationId should be Databases_ListBySqlServer. This will ensure consistency in the naming pattern across all the RPs in Azure and it will be easy for customers onbaording to Azure (reduces their learning curve).

RenaShahMSFT committed:
Merge pull request #1 from mmacy/780

Doc edits for import/export Swagger spec

RenaShahMSFT committed:
Addressed comments

AutorestCI commented:
No modification for Python

AutorestCI commented:
No modification for NodeJS

AutorestCI commented:
No modification for Ruby
